File: src/leetcode/first/Code_113_PathSUm2.java
package leetcode.first;

import leetcode.tool.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * 路径总和2
 *
 * @since 2020-5-24 Sunday 19:59 - 21:00
 */
public class Code_113_PathSUm2 {
    static List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;
        dfs(root, sum, res, new ArrayList<>());
        return res;
    }

    static void dfs(TreeNode node, int sum, List<List<Integer>> res, List<Integer> tmp) {
        List<Integer> list = new ArrayList<>(tmp);
        if (node == null) {
            return;
        } else if (node.left == null && node.right == null && (sum - node.val == 0)) {
            list.add(node.val);
            res.add(list);
            return;
        }
        list.add(node.val);
        dfs(node.left, sum - node.val, res, list);
        dfs(node.right, sum - node.val, res, list);

//        if (node == null && sum == 0 && !res.contains(tmp)) {
//            res.add(new ArrayList<>(tmp));
//            return;
//        }
//        if (node == null) {
//            return;
//        }
//        List<Integer> list = new ArrayList<>(tmp);
//        if ((sum - node.val) != 0) {
//            list.add(node.val);
//            if (node.left != null && node.right == null) {
//                dfs(node.left, sum - node.val, res, list);
//            } else if (node.right != null && node.left == null) {
//                dfs(node.right, sum - node.val, res, list);
//            } else {
//                dfs(node.left, sum - node.val, res, list);
//                dfs(node.right, sum - node.val, res, list);
//            }
//        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(11);
        root.right.left = new TreeNode(13);
        root.right.right = new TreeNode(4);
        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);
        root.right.right.left = new TreeNode(5);
        root.right.right.right = new TreeNode(1);
        System.out.println(pathSum(root, 22));
//        TreeNode root = new TreeNode(1);
//        root.right = new TreeNode(2);
//        System.out.println(pathSum(root, 1));
    }
}

File: src/c-leetcode/first/Code_113_PathSum2.c
/* 
   @since 2020-9-4 Friday
   @update 2020-9-7 Monday
   @finished 2020-9-8 Tuesday
 */

#include <stdio.h>
#include <stdlib.h>
#include "../TreeNode.h"

void dfs(struct TreeNode *root, int sum, int *tmp, int index, int *returnSize, int **returnColSizes, int **res)
{
    if (root == 0)
        return;
    if (root->left == 0 && root->right == 0 && sum - root->val == 0)
    {
        tmp[index++] = root->val;
        res[*returnSize] = malloc(index * sizeof(int));
        for (int i = 0; i < index; i++)
            res[*returnSize][i] = tmp[i];
        returnColSizes[0][*returnSize] = index;
        *returnSize += 1;
        return;
    }
    tmp[index++] = root->val;
    dfs(root->left, sum - root->val, tmp, index, returnSize, returnColSizes, res);
    dfs(root->right, sum - root->val, tmp, index, returnSize, returnColSizes, res);
}

int **pathSum(struct TreeNode *root, int sum, int *returnSize, int **returnColumnSizes)
{
    int **res = malloc(1024 * sizeof(int *));
    *returnColumnSizes = malloc(1024 * sizeof(int));
    if (res == 0 || returnColumnSizes == 0)
        exit(-1);
    *returnSize = 0;
    if (root == 0)
        return res;
    int *arr = malloc(1024 * sizeof(int));
    dfs(root, sum, arr, 0, returnSize, returnColumnSizes, res);
    printf("returnSize: %d\n", *returnSize);
    return res;
}

int main(int argc, char const *argv[])
{
    struct TreeNode *root = binaryTree();
    root->val = 1;
    int returnSize = 0;
    int **returnColumnSizes = malloc(100 * sizeof(int *));
    int **res = pathSum(root, 8, &returnSize, returnColumnSizes);
    printf("[\n");
    for (int i = 0; i < returnSize; i++)
    {
        printf(" [ ");
        for (int j = 0; j < returnColumnSizes[0][i]; j++)
            printf("%d ", res[i][j]);
        printf("]\n");
    }
    printf("]\n");
}
